src/haar.jl

```diff
@@ -0,0 +1,159 @@
+
+## References
+
+M. Oren, C. Papageorgiou, P. Sinha, E. Osuna and T. Poggio, "Pedestrian detection using wavelet templates," Proceedings of IEEE Computer Society Conference on Computer Vision and Pattern Recognition, San Juan, 1997, pp. 193-199.
+
+P. Viola and M. Jones, "Rapid object detection using a boosted cascade of simple features," Proceedings of the 2001 IEEE Computer Society Conference on Computer Vision and Pattern Recognition. CVPR 2001, 2001, pp. I-511-I-518 vol.1.
+
+"""
+function haar_features(img::AbstractArray{T, 2}, top_left::Array{I, 1}, bottom_right::Array{I, 1}, feat::Symbol, coordinates = nothing) where {T <: Union{Real, Color}, I <: Int}
+    check_coordinates(top_left, bottom_right)
+    nrect = check_feature_type(feat)
+
+    if coordinates == nothing
+        rectangle_features = haar_coordinates(bottom_right[1] - top_left[1] + 1, bottom_right[2] - top_left[2] + 1, feat)
+    else
+        a = Array{SVector{4,Int}}(nrect)
+        for i = 1:nrect
+            a[i] = SVector{4}(0,0,0,0)
+        end
+        rectangle_features = Array([a])
+        for i = 1:size(coordinates)[1]
+            temp = []
+            for j = 1:size(coordinates)[2]
+                push!(temp, SVector{4}(coordinates[i,j,1], coordinates[i,j,2], coordinates[i,j,3], coordinates[i,j,4]))
+            end
+            push!(rectangle_features, temp)
+        end
+        deleteat!(rectangle_features, 1)
+    end
+    if length(rectangle_features) > 0
+        feats = length(rectangle_features)
+        rects = length(rectangle_features[1])
+    else
+        println("No features of given type found in region!")
+        feats = 0
+        rects = 0
+    end
+
+    features = zeros(T, feats, rects)
+    for i = 1:feats
+        for j = 1:rects
+            features[i,j] = boxdiff(img, rectangle_features[i][j][1]:rectangle_features[i][j][3], rectangle_features[i][j][2]:rectangle_features[i][j][4])
+        end
+    end
+
+    output = (sum(features[:, 2:2:end], 2) - sum(features[:, 1:2:end], 2))[:]
+end
+
+"""
+```
+coordinates = haar_coordinates(height, width, feat)
+```
+
+Returns an array containing the coordinates of the Haar-like features of the specified type.
+
+Parameters:
+
+ -  height        = Height of the neighbourhood/window where the coordinates are computed
+ -  width         = Width of the neighbourhood/window where the coordinates are computed
+ -  feat          = A symbol specifying the type of the Haar-like feature to be found
+
+```
+    Currently, feat can take 5 values:
+
+    :x2 = Two rectangles along horizontal axis
+    :y2 = Two rectangles along vertical axis
+    :x3 = Three rectangles along horizontal axis
+    :y3 = Three rectangles along vertical axis
+    :xy4 = Four rectangles along horizontal and vertical axes
+```
+
+"""
+function haar_coordinates(h::Int, w::Int, feat::Symbol)
```

Comment:
Tried to look up our conversation about this code on Slack, but it looks like it has disappeared into the ether. From what I recall, I think we agreed that it would be worthwhile to add a more detailed explanation to the documentation of this function. We wanted to decide on what precisely we consider a `feature`, so that we can choose consistent names for the variables that act as containers of the `features`. 

It is also worth further clarifying the output of this function, and the output of the ` haar_features` function as well. 

The caller of the `haar_coordinates` function specifies a rectangular region in the image and a particular feature type, e.g. two rectangles side by side.  This function then computes the coordinates for the particular rectangle configuration for all possible locations, and all possible rectangle width & heights. 
For example, for a given rectangle size one effectively slides the rectangles across the entire image. 

```julia
+----------------------------------+
|                                  |
| +-----++-----+                   |
| |     ||     |                   |
| |     ||     |  +---------->     |
| +-----++-----+                   |
|                                  |
|       +                          |
|       |                          |
|       |                          |
|       |                          |
|       |                          |
|       |                          |
|       |                          |
|       |                          |
|       |                          |
|       v                          |
|                                  |
|                                  |
+----------------------------------+
```